gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -179,7 +193,11 @@ private Object doResumableUpload() throws IOException {
       int retriesAttempted = 0;
       Hasher objectHasher = Hashing.crc32c().newHasher();
       do {
-        responseObserver = new InsertChunkResponseObserver(uploadId, writeOffset, objectHasher);
+        // Holds list of most recent number of numberOfRequestsToRetain requests, so upload can be
+        // rewound and re-sent upon non-transient errors.
+        List<InsertObjectRequest.Builder> requestList = new ArrayList<>();
```

Comment:
I realize this came from my original implementation but: ArrayList.remove is inefficient - requires copying elements. How about switching to a LinkedList?

Reply:
There are two commits on this PR. You must be looking at first commit not all commits. I actually figured that TreeMap will be a good fit here and updated the code accordingly.